File: nodejs/binary-gap-868.js
/**
 * @param {number} N
 * @return {number}
 */
'use strict';
var binaryGap = function(N) {
    if ((N&(N-1))===0) {
        return 0;
    }
    let count = 0;
    let max = -1;
    let first = -1;
    while(N!==0) {
        if (N&1===1) {
            if (first===-1) {                
                first = count;
            } else {
                max = Math.max(max,count-first);
                first = count;
            }
            
        }
        count++;
        N>>>=1;        
    }
    return max;
};

console.log(binaryGap(1));
console.log(binaryGap(2));
console.log(binaryGap(4));
console.log(binaryGap(8));
console.log(binaryGap(22));
console.log(binaryGap(5));
console.log(binaryGap(6));
console.log(binaryGap(8));
console.log(binaryGap(536870912));
console.log(binaryGap(536870913));

File: golang/binarygap868/binary-gap-868.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println("HELLOOO")
	fmt.Println(binaryGap(536870912))
	fmt.Println(binaryGap(536870911))
}

func binaryGap(N int) int {
	if 0 == (N & (N - 1)) {
		return 0
	} else {
		first := -1
		count := 0
		max := -1
		for N != 0 {
			if N&1 == 1 {
				if first == -1 {
					first = count
				} else {
					max = m(max, count-first)
					first = count
				}
			}
			N >>= 1
			count++
		}
		return max
	}
}

func m(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}
